### ISSUE TITLE
Should throw AuthenticationFailure RedisConnectionException when redis password not correct

### ISSUE THREAD

**Issue by WeihanLi** [StackExchange/StackExchange.Redis#2386]
Hi there, the library did not throw authenticationFailure exception, it throws `RedisTimeoutException` instead like follows:

```
StackExchange.Redis.RedisTimeoutException: Timeout performing GET (5000ms), inst: 10, qu: 0, qs: 0, aw: False, bw: CheckingForTimeout, last-in: 0, cur-in: 0, sync-ops: 1, async-ops: 0, serverEndpoint: 127.0.0.1:6382, conn-sec: n
/a, mc: 1/1/0, mgr: 10 of 10 available, clientName: PXPDWDWLI(SE.Redis-v2.6.96.30123), IOCP: (Busy=0,Free=1000,Min=12,Max=1000), WORKER: (Busy=0,Free=32767,Min=12,Max=32767), POOL: (Threads=11,QueuedItems=0,CompletedItems=109),
v: 2.6.96.30123 (Please take a look at this article for some common client-side issues that can cause timeouts: https://stackexchange.github.io/StackExchange.Redis/Timeouts)
   at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server, T defaultValue) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1988
   at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server, T defaultValue) in /_/src/StackExchange.Redis/RedisBase.cs:line 62
   at StackExchange.Redis.RedisDatabase.StringGet(RedisKey key, CommandFlags flags) in /_/src/StackExchange.Redis/RedisDatabase.cs:line 2996

```

Version: `2.6.96`

Run up a redis with docker:

``` sh
docker run -d --name redis-server-with-pass -p 6382:6379  --restart=always redis:7.0-alpine -- requirepass "123456"
```

Test code as below:

``` c#
public static void RawRedisTest()
{
    try
    {
        var redis = SetupConnection("127.0.0.1:6382", "password");
        while (true)
        {
            try
            {
                redis.GetDatabase().StringGet("123");
            }
            catch (RedisConnectionException e)
            {
                Console.WriteLine(e.FailureType);
            }
            
            Thread.Sleep(3000);
        }
    }
    catch (Exception e)
    {
        Console.WriteLine(e);
    }
    
    
}

private static ConnectionMultiplexer SetupConnection(string host, string? password = null)
{
    var configOptions = new ConfigurationOptions();
    configOptions.EndPoints.Add(host);
    configOptions.Password = password;
    configOptions.ConnectTimeout = 5000;
    configOptions.SyncTimeout = 5000;
    configOptions.ReconnectRetryPolicy = new LinearRetry(1 * 1000);
    configOptions.AbortOnConnectFail = false;
    return ConnectionMultiplexer.Connect(configOptions);
}
```

**Comment by NickCraver** [comment]
The problem here is you're setting `AbortOnConnectFail` which will try to connect indefinitely, so each command is timing out waiting for a valid connection. To see the real error connecting, disable that to get an immediate error result.

We are going to change this up though - it's something we've been thinking through (not as simple as it seems), actually one of my main topics for a sync call today, for the case where the user has _never_ connected I'm going to tweak what we're showing for exceptions because this is being hit so often with people using that setting.

**Comment by WeihanLi** [comment]
@NickCraver great thanks for the reply, we use `AbortOnConnectFail` to reconnect the Redis server in case of the Redis server restarts or network issues, if `AbortOnConnectFail` is set to `true`, would it keep reconnecting the Redis server according to the `ReconnectRetryPolicy`? If then, when should we consider using the `AbortOnConnectFail` option? Thanks.

**Comment by NickCraver** [comment]
@WeihanLi That argument only affects the _initial_ connection, the scenario being your application comes up before the Redis server/firewall/whatever does especially in cloud deploys with big orchestrations. Or you don't want a momentary outage to block you from coming up. In _all_ cases we'll retry _re_-connecting to the server, yep.

**Comment by WeihanLi** [comment]
@NickCraver thanks very much, it really helps, we may also want to reconnect to the Redis server during the initial connection when the Redis server is down. 
What's the recommended config? What if we set `connectRetry` to `int.MaxValue` and `AbortOnConnectFail` to `false`, would it similar to `AbortOnConnectFail` with `true`?

**Comment by NickCraver** [comment]
@WeihanLi recommendations wouldn't change. If you set connectRetry sky-high, you're not going to come back from the Connect method until it completes successfully, so you wouldn't even have a reference to a ConnectionMultiplexer to do anything with. That's generally not a useful scenario for anything unless your explicit goal is to retry connecting forever and not proceed until that's done. A small connectRetry (even lowering it to 1) and setting AbortOnConnectFail is the scenario most people are after the startup-race environments like remote or large deploys so that the multiplexer is spun up and useful as soon as it's useful, e.g. checking for .IsConnected if you need to, rather than structuring your entire application around whether it's even been created or not.

**Comment by NickCraver** [comment]
Heads up: these changes to exception messages landed in 2.6.104, so closing this one out :)
